File: heap(priority queue)/total_cost_to_hire_k_workers.java
import java.util.PriorityQueue;

public class total_cost_to_hire_k_workers {
    public long totalCost(int[] costs, int k, int candidates) {
        int i = 0, j = costs.length-1;
        long res = 0;
        PriorityQueue<Integer> first_candidates = new PriorityQueue<>();
        PriorityQueue<Integer> last_candidates = new PriorityQueue<>();
        
        while(k > 0){
            while(first_candidates.size() < candidates && i <= j){
                first_candidates.add(costs[i++]);
            }
            while(last_candidates.size() < candidates && i <= j){
                last_candidates.add(costs[j--]);
            }

            
            int first_value = first_candidates.size()>0 ? first_candidates.peek() : Integer.MAX_VALUE;
            int last_value = last_candidates.size()>0 ? last_candidates.peek() : Integer.MAX_VALUE;            

            if(first_value <= last_value){
                res = res + first_value;
                first_candidates.poll();
            }
            else{
                res = res + last_value;
                last_candidates.poll();
            }
            k--;
        }
        return res;
    }
}

File: heap(priority queue)/total_cost_to_hire_k_workers.py
import heapq
from typing import List


class Solution:
    def totalCost(self, costs: List[int], k: int, candidates: int) -> int:
        i, j = 0, len(costs)-1
        first_candidates = []
        last_candidates = []
        res = 0
        while k > 0:
            while len(first_candidates) < candidates and i <= j:
                heapq.heappush(first_candidates, costs[i])
                i += 1
            while len(last_candidates) < candidates and i <= j:
                heapq.heappush(last_candidates, costs[j])
                j -= 1
            
            first_value = first_candidates[0] if first_candidates else float('infinity')
            last_value = last_candidates[0] if last_candidates else float('infinity')

            if first_value <= last_value:
                res += first_value
                heapq.heappop(first_candidates)
            else:
                res += last_value
                heapq.heappop(last_candidates)
            k -= 1
        return res